README.md
# collectd-cdn_fastly

Fastly CDN plugin for collectd.

## Configuration

```xml
<LoadPlugin python>
  Globals true
</LoadPlugin>

<Plugin python>
  ModulePath "/usr/lib/collectd/python"
  Import "cdn_fastly"

  <Module cdn_fastly>
    ApiKey "68b329da9893e34099c7d8ad5cb9c940"

    <Service "www">
      Id "6IqS8vK4QRMAlb1ByyjrJF"
    </Service>
    <Service "assets">
      Id "qd8G3pOP2nGw0UlSE04t8v
    </Service>
  </Module>
</Plugin>
```


CHANGELOG
2015-02-19 Release 0.2.1
- Bugfix: null values returned for hit_ratio should not error

2014-07-09 Release 0.2.0
- Use HTTPretty mock server for tests
- Explicitly catch non-200 responses from the API
- Support username/password authentication in addition to API key
- Force a new cookie if the existing one expires within 15s
- Drop existing cookie if new configuration is passed to the same object

2014-01-23 Release 0.1.0
- Public release to PyPI.
- Add tests.
- Fix bug detecting no services configured.
- Reset services dict when reconfiguring.
- Change service configuration syntax.

2013-11-25 Release 0.0.1
- Internal use only.


test/helpers.py
from nose.tools import *
from mock import MagicMock, patch, call

import os
import Cookie
import datetime
import httpretty

HERE = os.path.dirname(__file__)

def fixture(name):
    return open(os.path.join(HERE, 'fixtures', name)).read()

def fastly_cookie(value, period=300):
    c = Cookie.SimpleCookie()
    c['fastly.session'] = value
    sc = c['fastly.session']

    expires = datetime.datetime.utcnow() + datetime.timedelta(seconds=period)
    sc['Expires'] = expires.strftime("%a, %d %b %Y %H:%M:%S GMT")
    sc['Path'] = '/'
    sc['secure'] = True
    sc['HttpOnly'] = True

    return sc.OutputString()

class CollectdConfig(object):
    """
    Quacks like collectd Config object.
    """
    def __init__(self, key, vals, children):
        self.key = key
        self.values = (vals,)
        self.children = [
            CollectdConfig(k,v,c)
            for k,v,c in children
        ]


test/unit/test_fastly.py
from ..helpers import *

import json
import copy
import datetime

class TestFastly(object):
    def setup(self):
        self.collectd = MagicMock()
        self.modules = patch.dict('sys.modules', {'collectd': self.collectd})
        self.modules.start()

        from collectd_cdn import fastly
        self.fastly = fastly.CdnFastly()

    def teardown(self):
        self.modules.stop()


class TestConfig(TestFastly):
    def test_all_valid_ints_as_strings(self):
        config = CollectdConfig('root', (), (
            ('ApiKey', 'abc123', ()),
            ('ApiTimeout', '13', ()),
            ('DelayMins', '17', ()),
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
            ('Service', (), (
                ('Name', 'two', ()),
                ('Id', '222', ()),
            )),
        ))
        self.fastly.config(config)

        assert_equal(self.fastly.api_key, 'abc123')
        assert_equal(self.fastly.api_timeout, 13)
        assert_equal(self.fastly.delay_mins, 17)
        assert_equal(self.fastly.services, {
            'one': '111',
            'two': '222',
        })

    def test_service_reconfig(self):
        config = CollectdConfig('root', (), (
            ('ApiKey', 'abc123', ()),
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
        ))
        self.fastly.config(config)
        assert_equal(self.fastly.services, { 'one': '111' })

        config = CollectdConfig('root', (), (
            ('ApiKey', 'abc123', ()),
            ('Service', (), (
                ('Name', 'two', ()),
                ('Id', '222', ()),
            )),
        ))
        self.fastly.config(config)
        assert_equal(self.fastly.services, { 'two': '222' })

    def test_session_reconfig(self):
        config = CollectdConfig('root', (), (
            ('ApiKey', 'abc123', ()),
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
        ))
        self.fastly.config(config)
        self.fastly.session.headers.update({'abc': 123})

        assert_equal(self.fastly.session.headers.get('abc'), 123)
        self.fastly.config(config)
        assert_equal(self.fastly.session.headers.get('abc'), None)

    def test_apikey(self):
        config = CollectdConfig('root', (), (
            ('ApiKey', 'abc123', ()),
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
        ))
        self.fastly.config(config)

        assert_equal(self.fastly.api_key, 'abc123')

    def test_apiuser_and_apipass(self):
        config = CollectdConfig('root', (), (
            ('ApiUser', 'abc', ()),
            ('ApiPass', '123', ()),
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
        ))
        self.fastly.config(config)

        assert_equal(self.fastly.api_user, 'abc')
        assert_equal(self.fastly.api_pass, '123')

    def test_apiuser_no_apipass(self):
        config = CollectdConfig('root', (), (
            ('ApiUser', 'abc', ()),
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
        ))
        assert_raises(Exception, self.fastly.config, config)

    def test_apipass_no_apiuser(self):
        config = CollectdConfig('root', (), (
            ('ApiPass', '123', ()),
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
        ))
        assert_raises(Exception, self.fastly.config, config)

    def test_no_credentials(self):
        config = CollectdConfig('root', (), (
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
        ))
        assert_raises(Exception, self.fastly.config, config)

    def test_all_credentials(self):
        config = CollectdConfig('root', (), (
            ('ApiKey', 'abc123', ()),
            ('ApiUser', 'abc', ()),
            ('ApiPass', '123', ()),
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
        ))
        assert_raises(Exception, self.fastly.config, config)

    def test_no_services(self):
        config = CollectdConfig('root', (), (
            ('ApiKey', 'abc123', ()),
        ))
        assert_raises(Exception, self.fastly.config, config)

    @patch('collectd_cdn.fastly.collectd.warning')
    def test_unknown_key(self, warning_mock):
        config = CollectdConfig('root', (), (
            ('Zebra', 'stripes', ()),
            ('ApiKey', 'abc123', ()),
            ('Service', (), (
                ('Name', 'one', ()),
                ('Id', '111', ()),
            )),
        ))
        self.fastly.config(config)
        warning_mock.assert_called_with("cdn_fastly plugin: Unknown config key: Zebra")


class TestScaleAndType(TestFastly):
    def test_time(self):
        v, t = self.fastly.scale_and_type('hits_time', 946.1020896459992)
        assert_equal(v, 946.1020896459992)
        assert_equal(t, 'response_time')

    def test_ratio(self):
        v, t = self.fastly.scale_and_type('hit_ratio', '0.9836E0')
        assert_equal(v, 0.9836)
        assert_equal(t, 'cache_ratio')

    def test_ratio_null(self):
        v, t = self.fastly.scale_and_type('hit_ratio', None)
        assert_equal(v, 0.0)
        assert_equal(t, 'cache_ratio')

    def test_size(self):
        v, t = self.fastly.scale_and_type('body_size', 219004331934)
        # FIXME: Should be float 3650072198.9 ?
        assert_equal(v, 3650072198)
        assert_equal(t, 'bytes')

    def test_other(self):
        v, t = self.fastly.scale_and_type('status_2xx', 11152796)
        # FIXME: Should be float 185879.93333333332 ?
        assert_equal(v, 185879)
        assert_equal(t, 'requests')


class TestGetTimeRange(TestFastly):
    @patch('collectd_cdn.fastly.CdnFastly._now')
    def test_delay(self, now_mock):
        self.fastly.delay_mins = 30

        now_mock.return_value = datetime.datetime(2014, 2, 1, 12, 30, 0, 0)
        t_from, t_to = self.fastly.get_time_range()
        t_to = datetime.datetime.fromtimestamp(t_to)
        t_from = datetime.datetime.fromtimestamp(t_from)

        assert_equal(t_to, datetime.datetime(2014, 2, 1, 12, 0, 0, 0))
        assert_equal(t_from, datetime.datetime(2014, 2, 1, 11, 59, 0, 0))

    @patch('collectd_cdn.fastly.CdnFastly._now')
    def test_round_down(self, now_mock):
        now_mock.return_value = datetime.datetime(2014, 2, 1, 12, 13, 14, 15)
        t_from, t_to = self.fastly.get_time_range()
        t_to = datetime.datetime.fromtimestamp(t_to)
        t_from = datetime.datetime.fromtimestamp(t_from)

        assert_equal(t_to, datetime.datetime(2014, 2, 1, 12, 3, 0, 0))
        assert_equal(t_from, datetime.datetime(2014, 2, 1, 12, 2, 0, 0))


class TestRequest(TestFastly):
    def __init__(self):
        self.MOCK_LOGIN_URL = "https://api.fastly.com/login"
        self.MOCK_STATS_URL = "https://api.fastly.com/stats/service/mocked"

    @httpretty.activate
    def test_request_invalid_response(self):
        httpretty.register_uri(
            httpretty.GET,
            self.MOCK_STATS_URL,
            body='',
            status=500
        )

        assert_raises(Exception, self.fastly.request, 'mocked', 1, 2)

    @httpretty.activate
    def test_request_service_and_range(self):
        httpretty.register_uri(
            httpretty.GET,
            self.MOCK_STATS_URL,
            body='{"data": {}}'
        )

        self.fastly.api_key = 'abc123'
        self.fastly.request('mocked', 1390320360, 1390320420)

        assert_equal(httpretty.last_request().querystring, {
            'by': ['minute'],
            'to': ['1390320420'],
            'from': ['1390320360'],
        })

    @httpretty.activate
    def test_request_api_key(self):
        httpretty.register_uri(
            httpretty.GET,
            self.MOCK_STATS_URL,
            body='{"data": {}}'
        )

        self.fastly.api_key = 'abc123'
        self.fastly.request('mocked', 1, 2)

        assert_equal(httpretty.last_request().headers.get('Fastly-Key'), 'abc123')

    @httpretty.activate
    def test_request_user_pass_login_once(self):
        cookie = fastly_cookie('just-one')
        expected_cookie = 'fastly.session=just-one'

        httpretty.register_uri(
            httpretty.POST,
            self.MOCK_LOGIN_URL,
            responses=[
                httpretty.Response(
                    body='{"user": {}}', status=200,
                    adding_headers={ 'Set-Cookie': cookie }),
                httpretty.Response(
                    body='should not be called',
                    status=400),
            ]
        )
        httpretty.register_uri(
            httpretty.GET,
            self.MOCK_STATS_URL,
            body='{"data": {}}'
        )

        self.fastly.api_user = 'abc'
        self.fastly.api_pass = '123'
        self.fastly.request('mocked', 1, 2)
        assert_equal(httpretty.last_request().headers.get('Cookie'), expected_cookie)

        self.fastly.request('mocked', 3, 4)
        assert_equal(httpretty.last_request().headers.get('Cookie'), expected_cookie)

    @httpretty.activate
    def test_request_user_pass_cookie_expiring_15secs(self):
        short_cookie = fastly_cookie('short', 10)
        long_cookie  = fastly_cookie('long', 30)

        httpretty.register_uri(
            httpretty.POST,
            self.MOCK_LOGIN_URL,
            responses=[
                httpretty.Response(
                    body='{"user": {}}', status=200,
                    adding_headers={ 'Set-Cookie': short_cookie }),
                httpretty.Response(
                    body='{"user": {}}', status=200,
                    adding_headers={ 'Set-Cookie': long_cookie }),
                httpretty.Response(
                    body='should not be called',
                    status=400),
            ]
        )
        httpretty.register_uri(
            httpretty.GET,
            self.MOCK_STATS_URL,
            body='{"data": {}}'
        )

        self.fastly.api_user = 'abc'
        self.fastly.api_pass = '123'
        self.fastly.request('mocked', 1, 2)
        assert_equal(
            httpretty.last_request().headers.get('Cookie'),
            'fastly.session=short'
        )

        self.fastly.request('mocked', 3, 4)
        assert_equal(
            httpretty.last_request().headers.get('Cookie'),
            'fastly.session=long'
        )

        self.fastly.request('mocked', 5, 6)
        assert_equal(
            httpretty.last_request().headers.get('Cookie'),
            'fastly.session=long'
        )

    @httpretty.activate
    def test_response_json(self):
        fixture_json = fixture('api_response.json')
        fixture_data = json.loads(fixture_json)['data']

        httpretty.register_uri(
            httpretty.GET,
            self.MOCK_STATS_URL,
            body=fixture_json
        )

        self.fastly.api_key = 'abc123'
        t_from, t_to = self.fastly.get_time_range()
        resp_json = self.fastly.request('mocked', t_from, t_to)

        assert_equal(resp_json, fixture_data)


class TestRead(TestFastly):
    @patch('collectd_cdn.fastly.CdnFastly.get_time_range')
    @patch('collectd_cdn.fastly.CdnFastly.request')
    @patch('collectd_cdn.fastly.CdnFastly.submit')
    @patch('collectd_cdn.fastly.collectd.warning')
    def test_three_services_one_error(self, warn_mock, submit_mock, req_mock, range_mock):
        self.fastly.services = {
            'one': '111',
            'two': '222',
            'three': '333',
        }

        range_mock.return_value = (1390320360, 1390320420)
        fixture_data = json.loads(fixture('simple_service.json'))
        req_mock.side_effect = [
            copy.deepcopy(fixture_data),
            ValueError("No JSON object could be decoded"),
            copy.deepcopy(fixture_data),
        ]

        self.fastly.read()

        req_calls = [
            call('111', *range_mock),
            call('222', *range_mock),
            call('333', *range_mock),
        ]
        submit_calls = [
            call('three', 'hits', 'requests', 777, 1390320360),
            call('three', 'hits_time', 'response_time', 3.5722524239999993, 1390320360),
            call('one', 'hits', 'requests', 777, 1390320360),
            call('one', 'hits_time', 'response_time', 3.5722524239999993, 1390320360),
        ]

        assert_equal(req_mock.call_count, len(req_calls))
        submit_mock.assert_has_calls(submit_calls)
        assert_equal(submit_mock.call_count, len(submit_calls))
        warn_mock.assert_called_with("cdn_fastly plugin: Failed to query service: two")


collectd_cdn/fastly.py
#!/usr/bin/env python
"""
Read plugin for Fastly CDN stats.

TODO:
    - Track last query date so that we can fill in blanks?
"""

import collectd
import json
import time
import requests
import cookielib
import datetime
import calendar

# Must be rounded to 1min because that's the highest resolution that Fastly
# provide data for. collectd will only call the plugin this often.
# Ref: http://docs.fastly.com/api/stats#Range
INTERVAL = 60

class CdnFastly(object):
    def __init__(self):
        self.LOGIN_URL = "https://api.fastly.com/login"
        self.STATS_URL = "https://api.fastly.com/stats/service/%(service_id)s"
        self.PLUGIN_NAME = "cdn_fastly"
        self.COOKIE_EXPIRE = 15

        self.delay_mins = 10
        self.api_timeout = 5
        self.api_key = None
        self.api_user = None
        self.api_pass = None
        self.session = requests.Session()
        self.services = {}

    def _warn(self, message):
        collectd.warning("cdn_fastly plugin: %s" % message)

    def _raise(self, message):
        raise Exception("cdn_fastly plugin: %s" % message)

    def _now(self):
        return datetime.datetime.now()

    def config(self, conf):
        """
        Configure the plugin.
        """
        # Reset any previously configured services.
        self.services = {}
        # Reset session (incl. authentication)
        self.session = requests.Session()

        for node in conf.children:
            if node.key == 'ApiKey':
                self.api_key = node.values[0]
            elif node.key == 'ApiUser':
                self.api_user = node.values[0]
            elif node.key == 'ApiPass':
                self.api_pass = node.values[0]
            elif node.key == 'ApiTimeout':
                self.api_timeout = int(node.values[0])
            elif node.key == 'DelayMins':
                self.delay_mins = int(node.values[0])
            elif node.key == 'Service':
                s_name, s_id = (None, None)
                for s_node in node.children:
                    if s_node.key == 'Name':
                        s_name = s_node.values[0]
                    elif s_node.key == 'Id':
                        s_id = s_node.values[0]
                    else:
                        self._warn("Unknown config key: %s" % node.key)
                if not (s_name and s_id):
                    self._raise("Invalid 'Service' config")
                self.services[s_name] = s_id
            else:
                self._warn("Unknown config key: %s" % node.key)

        if not (self.api_key or (self.api_user and self.api_pass)):
            self._raise("No ApiKey or ApiUser/ApiPass configured")

        if (self.api_key and (self.api_user or self.api_pass)):
            self._raise("ApiKey and ApiUser/ApiPass are mutually exclusive")

        if len(self.services) < 1:
            self._raise("No Service blocks configured")

    def read(self):
        """
        Fetch and submit data. Called once per INTERVAL.
        """
        time_from, time_to = self.get_time_range()

        for service_name, service_id in self.services.items():
            try:
                service_data = self.request(service_id, time_from, time_to)
            except:
                self._warn("Failed to query service: %s" % service_name)
                continue

            for service_period in service_data:
                vtime = service_period.pop('start_time')
                del service_period['service_id']

                for key, val in service_period.items():
                    val, vtype = self.scale_and_type(key, val)
                    self.submit(service_name, key, vtype, val, vtime)

    def scale_and_type(self, key, val):
        """
        Find the appropriate data type and scale the value by INTERVAL
        where required.

        Ref: http://collectd.org/documentation/manpages/types.db.5.shtml
        """
        if key.endswith('_time'):
            vtype = 'response_time'
        elif key.endswith('_ratio'):
            vtype = 'cache_ratio'
            if val == None:
                val = 0
            val = float(val)
        elif key.endswith('_size') or key == 'bandwidth':
            vtype = 'bytes'
            val = val / INTERVAL
        else:
            vtype = 'requests'
            val = val / INTERVAL

        return val, vtype

    def get_time_range(self):
        """
        Construct a time range for which to query stats for.

        This is called once per `self.read()` so that we query the same time
        period for all services consistently no matter how long the request
        takes.

        A delay of `self.delay_mins` is applied because Fastly's data from
        edge is 10~15 mins behind the present time.

        Ref: http://docs.fastly.com/api/stats#Availability
        """
        # Timestamp rounded down to the minute.
        now = calendar.timegm(
            self._now().replace(
                second=0, microsecond=0
            ).utctimetuple()
        )

        time_to = now - (self.delay_mins * 60)
        time_from = time_to - INTERVAL

        return time_from, time_to

    def submit(self, service_name, metric_name, metric_type, value, time):
        """
        Submit a single metric with the appropriate properties.
        """
        v = collectd.Values()
        v.plugin = self.PLUGIN_NAME
        v.plugin_instance = service_name

        v.type = metric_type
        v.type_instance = metric_name

        v.time = time
        v.values = [value, ]
        v.interval = INTERVAL

        v.dispatch()

    def auth(self):
        """
        Setup authentication headers or cookies for the session.
        """
        if self.api_key:
            self.session.headers.update({
                'Fastly-Key': self.api_key,
            })

            return

        if self.api_user and self.api_pass:
            # Force a new cookie if it's going to expire soon.
            valid_until = time.time() + self.COOKIE_EXPIRE
            valid_cookies = [
                c for c in self.session.cookies
                if not c.is_expired(valid_until)
            ]

            if len(valid_cookies) < 1:
                payload = {
                    'user': self.api_user,
                    'password': self.api_pass,
                }
                resp = self.session.post(
                    self.LOGIN_URL,
                    data=payload,
                    timeout=self.api_timeout
                )
                if resp.status_code != 200:
                    self._raise("Non-200 response from /login")

            return

        self._raise("No authentication methods configured")

    def request(self, service_id, time_from, time_to):
        """
        Requests stats from Fastly's API and return a dict of data. May
        contain multiple time periods.
        """
        params = {
            'from': time_from,
            'to': time_to,
            'by': "minute",
        }
        url = self.STATS_URL % {
            'service_id': service_id,
        }
        headers = {
            'Fastly-Key': self.api_key,
        }

        self.auth()
        resp = self.session.get(
            url,
            params=params,
            timeout=self.api_timeout
        )
        if resp.status_code != 200:
            self._raise("Non-200 response")

        data = resp.json()['data']
        return data

cdn_fastly = CdnFastly()
collectd.register_config(cdn_fastly.config)
collectd.register_read(cdn_fastly.read, INTERVAL)
